### PR TITLE
E2e test

### PR DESCRIPTION
**Pull request by commoddity**
Created Dockerized E2E tests for the entire stack required by Relay Meter and the Collector and API Server containers.

In future PRs should add test cases for covering error response from the API and error potentials in the Collector but I figured this PR is big enough already.

### CODE REVIEW

File: .github/workflows/build-test.yml
Comment (adshmh) [inline .github/workflows/build-test.yml:59]: Is this step needed, considering the makfile target on `run_e2e_tests` already has a dependency on `test_env_up`?

File: cmd/collector/main.go
Comment (adshmh) [inline cmd/collector/main.go:18]: Thank you for fixing the camelCase on these constants!
Not a blocker, but we will likely need a configuration data structure, defined in the collector's package, which is then here read from a single json-formatted configuration file. This would result in removing all these environment variables in favour of just one, which would point to the configuration file's location.

File: cmd/collector/main.go
Comment (adshmh) [inline cmd/collector/main.go:22]: Along the same line as previous comment: thank you for fixing the naming style, and these should eventually move to the `collecor` package itself and used by its methods, rather than defined and set here.

File: cmd/collector/main.go
Comment (adshmh) [inline cmd/collector/main.go]: I'm OK with change in data type, but perhaps the field name should change to `maxArchiveAge` instead?

File: cmd/collector/main.go
Comment (adshmh) [inline cmd/collector/main.go:49]: Could you clarify why this change is needed? Reporting errors using `Errorf` seems more reasonable.

File: db/influx.go
Comment (adshmh) [inline db/influx.go]: OK to move to a separate package IMO if this is going to be used in multiple places, but hopefully to a small, focused package with a name indicating what it does, rather than `utils-go`

File: docker-compose.test.yml
Comment (adshmh) [inline docker-compose.test.yml]: Could we use a specific image tag please? `latest` is not a great choice for production use.

File: tests/e2e_test.go
Comment (adshmh) [inline tests/e2e_test.go:3]: Would it make sense to use a build tag here? This only needs to be built during E2E tests.

File: tests/e2e_test.go
Comment (adshmh) [inline tests/e2e_test.go]: Would it make sense to verify the actual count per origin here?

File: tests/e2e_test.go
Comment (adshmh) [inline tests/e2e_test.go]: Since this repo is now maintained on pocket-foundation org, could we update the imports accordingly please?

File: tests/e2e_test.go
Comment (adshmh) [inline tests/e2e_test.go]: If I am reading this test function correctly, it is populating InfluxDB with records, and then reading data from InfluxDB using the `db` package's methods. This does not make it an E2E test IMO, but more of an integration test (and valuable of course). We can rename this and move it to a package specific to integration tests (with correct built tag). 
Alternatively, we could move the key parts, e.g. data population, to the subsequent E2E test and remove this altogether, as the API server E2E test covers the entire stack.

File: tests/e2e_test.go
Comment (adshmh) [inline tests/e2e_test.go]: I think this comment, and all subsequent similar comments, are signalling the need for breaking this function into multiple E2E test functions. Breaking apart this function will make the actual test more readable/maintainable. 
The data and the API server itself can be shared among all these tests through some builder function (of which you have multiple examples here already).

File: tests/e2e_test.go
Comment (adshmh) [inline tests/e2e_test.go]: Please see above comment regarding breaking this into multiple test functions: each E2E test function will then focus on a specific aspect of the relay-meter.

File: tests/e2e_testdata.go
Comment (adshmh) [inline tests/e2e_testdata.go]: From GO's standard `time` package examples, the following may read better:
```
func startOfDay(t time.Time) time.Time {
     return time.Date(t.Year(), t.Month(), t.Day(), 0, 0, 0, 0, t.Location())
}
```

File: tests/e2e_testdata.go
Comment (adshmh) [inline tests/e2e_testdata.go]: Technically this goes over to the start of the next day, is this the intended value (considering the name is `endOfDay`)?

File: tests/e2e_testdata.go
Comment (adshmh) [inline tests/e2e_testdata.go]: Would it make the code more readable if the data here, i.e. all the test relays, were in a json-formatted file that is read/unmarshalled into this array? This would allow updating the test data by editing the dataset in json, rather than modifying GO code.

File: tests/e2e_testdata.go
Comment (adshmh) [inline tests/e2e_testdata.go]: These strings MAY be more maintainable/readable if put into separate text files that are simply read here, to avoid embedding all of this into GO code (less certain about this though than with the test relays above)

File: tests/e2e_testdata.go
Comment (adshmh) [inline tests/e2e_testdata.go]: Perhaps all the above could be stored in a map instead to better represent their relationship, i.e. task name and task contents?

**Review by adshmh — CHANGES_REQUESTED** [review]
Thank you for this PR. It goes a long way towards adding the much-needed E2E tests to relay-meter.

File: .github/workflows/build-test.yml
Comment (commoddity) [reply .github/workflows/build-test.yml:59]: `run_e2e_tests` doesn't start the test env. `test_e2e` does but I use that for running locally. I figured keeping the steps separate for the CI file was clearer and easier to read.

File: cmd/collector/main.go
Comment (commoddity) [reply cmd/collector/main.go:18]: SG, I will add that in a separate PR later on.

File: cmd/collector/main.go
Comment (commoddity) [reply cmd/collector/main.go:22]: SG, same as above.

File: cmd/collector/main.go
Comment (commoddity) [reply cmd/collector/main.go]: I suppose it could but I still feel including days in the variable name makes it clear what the duration is vs having to check in the `gatherOptions` func to see what the time period being discussed is. ie. the duration is still a # of days even if its converted to a duration before being added to the struct.

File: cmd/collector/main.go
Comment (commoddity) [reply cmd/collector/main.go:49]: This was to conform with a linter error that I was getting in my editor.

Side note: we currently have `golang-ci-lint` disabled in both `pre-commit` and the CI as this repo has a whole ton of linter issues that need to be fixed to conform to the rest of the backend repos' linter style. Once this E2E PR is finished and merged I suggest we turn those linters back on and make all the changes.

File: db/influx.go
Comment (commoddity) [reply db/influx.go]: We already have a `time` subpackage inside `utils-go`  for time related funcs so I was thinking to put it there.

File: docker-compose.test.yml
Comment (commoddity) [reply docker-compose.test.yml]: Sure, I will find out which version of influx we use in production and use that tag here.

File: tests/e2e_test.go
Comment (commoddity) [reply tests/e2e_test.go:3]: Yes that's a good idea (just had to google build tags). I will add this to the E2E files.

File: tests/e2e_test.go
Comment (commoddity) [reply tests/e2e_test.go]: Unfortunately that's not really possible with the existing `influx.go` code since there is no explicit `stop` date in the Influx query, meaning Influx automatically fills it in with the current time so the # of relays per day will vary based on the when the `TodaysCountPerOrigin` Influx method gets called.

File: tests/e2e_test.go
Comment (commoddity) [reply tests/e2e_test.go]: Yes definitely, I will update.

File: tests/e2e_test.go
Comment (commoddity) [reply tests/e2e_test.go]: Sure, I can agree with that definition. My thought process was to populate the Influx DB and run the collector, then verify that the records actually existed in Influx before hitting the API Server endpoints (which is the part that satisfies the "E2E" definition). If the records don't exist in Influx then there's no point proceeding any farther. 

But you are correct that using the db methods alone does not constitute an E2E test. I don't mind moving them to an integration test elsewhere (as I think it's worthwhile to have coverage on those methods) and then only care about the AP Server results in the E2E.

File: tests/e2e_test.go
Comment (commoddity) [reply tests/e2e_test.go]: Sure I can refactor a bit  (as per the above comment) and then make each endpoint its own test.

File: tests/e2e_test.go
Comment (commoddity) [reply tests/e2e_test.go]: Same as above.

File: tests/e2e_testdata.go
Comment (commoddity) [reply tests/e2e_testdata.go]: Sure NP. I will also move it to the `utils-go` `time` package.

File: tests/e2e_testdata.go
Comment (commoddity) [reply tests/e2e_testdata.go]: Yes you are right. It goes to exactly midnight of the next day which is arguably the "end of the same day" but I can just subtract a ms from it to be exactly correct.

File: tests/e2e_testdata.go
Comment (commoddity) [reply tests/e2e_testdata.go:53]: I don't have a strong opinion about it either way. There are already JSON formatted strings in this Go file (necesary to subtitute variables into the string) so my preference is to have everything in one test data file rather than multiple files of different formats but ultimately I'm not strongly attached. WDYT?

File: tests/e2e_testdata.go
Comment (commoddity) [reply tests/e2e_testdata.go]: Again I am fine if you want to change it but my preference is everything in one file.

File: tests/e2e_testdata.go
Comment (commoddity) [reply tests/e2e_testdata.go]: Sure that makes sense, I will update.

File: tests/e2e_testdata.go
Comment (commoddity) [reply tests/e2e_testdata.go]: nm, I moved them to a JSON file

File: tests/e2e_testdata.go
Comment (commoddity) [reply tests/e2e_testdata.go]: moved them to a JSON file

File: tests/e2e_test.go
Comment (commoddity) [reply tests/e2e_test.go]: Done, I've refactored these tests into their own funcs

File: tests/e2e_test.go
Comment (commoddity) [reply tests/e2e_test.go]: Refactored

File: tests/e2e_test.go
Comment (commoddity) [reply tests/e2e_test.go]: This is done now.

File: tests/e2e_test.go
Comment (commoddity) [reply tests/e2e_test.go]: These tests are removed since the E2E covers them. Can potentially add an integration test to the collector in a later PR.

File: docker-compose.test.yml
Comment (commoddity) [reply docker-compose.test.yml]: Done

**Review by ricantar — DISMISSED** [review]
LG! Thanks you for adding and breaking those test

File: cmd/collector/main.go
Comment (adshmh) [reply cmd/collector/main.go]: The `maxArchiveAgeDays` itself is specified as a `Time.Duration` with this PR. The source of this field is `MAX_ARCHIVE_AGE`, which does not even have a `DAYS`prefix or suffix (perhaps it should?).
Perhaps in a subsequent PR, we can consolidate all these options into a single struct that is read from a json config file (see other repos for samples please).

File: cmd/collector/main.go
Comment (adshmh) [reply cmd/collector/main.go:49]: Sure, I agree we should handle any reasonable suggestions by a linter in a separate PR. Could you revert this change please? Then, as you sugested above, after merging this PR we can handle the linter issues.

File: tests/e2e_suite.go
Comment (adshmh) [inline tests/e2e_suite.go:384]: Thank you for the helpful comment on why this is needed here.

**Review by adshmh — CHANGES_REQUESTED** [review]
Thank you for the updates. This reads much better now.
Please just address the two minor comments from previous review so we can merge.

File: cmd/collector/main.go
Comment (commoddity) [reply cmd/collector/main.go]: OK, I've updated the var name.

File: cmd/collector/main.go
Comment (commoddity) [reply cmd/collector/main.go:49]: Unless I'm misunderstanding it, doesn't `fmt.Errorf` merely return a string and not log it? So in this case the existing code wasn't doing anything since the return value was never checked or used.

Comment (commoddity) [reply cmd/collector/main.go:49]: The linter error is `result of fmt.Errorf call not used`

Comment (adshmh) [reply cmd/collector/main.go:49]: I see. My mistake, somehow I read it as `t.Errorf`.

**Review by adshmh — APPROVED** [review]

**Comment by adshmh** [comment]
LGTM. But I think we should hold off on the merge until early next week.

**Comment by commoddity** [comment]
> LGTM. But I think we should hold off on the merge until early next week.

Sounds good to me; I will merge on Monday. Thank you for the review.

**Pull request merged**
